File: src/runtime/boxes/bb_fence.c
/* _XFNCE    FENCE       succeed once; β cuts (no retry) */
#include "bb_box.h"
#include <stdlib.h>
#include <string.h>


spec_t bb_fence(void *zeta, int entry)
{
    fence_t *ζ = zeta;
    if (entry==α)                                                               goto FENCE_α;
    if (entry==β)                                                               goto FENCE_β;
    FENCE_α:        ζ->fired=1;                                                 goto FENCE_γ;
    FENCE_β:                                                                    goto FENCE_ω;
    FENCE_γ:                                                                    return spec(Σ+Δ,0);
    FENCE_ω:                                                                    return spec_empty;
}

fence_t *bb_fence_new(void)
{ return calloc(1,sizeof(fence_t)); }

File: src/runtime/boxes/BbFence.java
package snobol4.runtime.boxes;

/**
 * BbFence.java — FENCE: succeed once; β cuts (no retry)
 * Port of bb_fence.c / bb_fence.s
 *
 *   FENCE_α:  fired=1;            goto FENCE_γ;
 *   FENCE_β:                      goto FENCE_ω;
 *   FENCE_γ:  return spec(Σ+Δ,0);
 *   FENCE_ω:  return spec_empty;
 */
public class BbFence extends BbBox {

    public BbFence(MatchState ms) { super(ms); }

    @Override public Spec alpha() { return new Spec(ms.delta, 0); }  // FENCE_γ

    @Override public Spec beta()  { return null; }                   // FENCE_ω
}
